linera-explorer/src/lib.rs

```diff
@@ -221,52 +230,56 @@ fn format_bytes(value: &JsValue) -> JsValue {
     modified_value
 }
 
+fn page_name_and_args(page: &Page) -> (&str, Vec<(String, String)>) {
+    match page {
+        Page::Unloaded | Page::Home { .. } => ("", Vec::new()),
+        Page::Block(b) => ("block", vec![("block".to_string(), b.hash.to_string())]),
+        Page::Blocks { .. } => ("blocks", Vec::new()),
+        Page::Applications(_) => ("applications", Vec::new()),
+        Page::Error(_) => ("error", Vec::new()),
+    }
+}
+
+fn find_arg(args: &[(String, String)], key: &str) -> Option<String> {
+    args.iter().find_map(|(k, v)| { if k == key { Some(v.clone()) } else { None }})
+}
+
+fn chain_id_from_args(app: &JsValue, data: &Data, args: &[(String, String)]) -> ChainId {
+    match find_arg(args, "chain") {
+        None => data.chain,
+        Some(chain_id) => {
+            let chain_js: JsValue = chain_id
+                .serialize(&SER)
+                .expect("failed to serialize ChainId");
+            setf(app, "chain", &chain_js);
+            ChainId::from_str(&chain_id).unwrap()
+        }
+    }
+}
+
 /// Main function to switch between vue.js pages
 async fn route_aux(
     app: &JsValue,
     data: &Data,
     path: &Option<String>,
-    chain_id: &ChainId,
     args: &[(String, String)],
 ) {
-    let chain_js: JsValue = chain_id
-        .serialize(&SER)
-        .expect("failed to serialize ChainId");
-    setf(app, "chain", &chain_js);
-    let path = match (path, data.page.clone()) {
-        (Some(p), _) => p,
-        (_, Page::Unloaded | Page::Home { .. }) => "",
-        (_, Page::Block(_)) => "block",
-        (_, Page::Blocks { .. }) => "blocks",
-        (_, Page::Applications(_)) => "applications",
-        (_, Page::Error(_)) => "error",
+    let chain_id = chain_id_from_args(app, data, args);
+    let (page_name, args): (&str, Vec<(String, String)>) = match (path, &data.page) {
+        (Some(p), _) => (p, args.to_vec()),
+        (_, p) => page_name_and_args(p)
     };
-    let address = node_service_address(&data.config, false);
-    let result = match path {
-        "" => home(&address, chain_id).await,
+    let address = node_service_address(&data.config, Protocol::Http);
+    let result = match page_name {
+        "" => home(&address, &chain_id).await,
         "block" => {
-            let hash = args.iter().find_map(|(k, v)| {
-                if k == "block" {
-                    Some(CryptoHash::from_str(v).unwrap())
-                } else {
-                    None
-                }
-            });
-            block(&address, chain_id, hash).await
+            let hash = find_arg(&args, "block").map(|hash| CryptoHash::from_str(&hash).unwrap());
```

Comment:
This `unwrap` can fail because the user entered a wrong URI, can it?
Maybe we should print an error that the block hash is invalid.